src/tests/utests/xi_utest_sft_logic.c

```diff
@@ -231,6 +231,292 @@ XI_TT_TESTCASE_WITH_SETUP(
+                           .msgver  = 1},
+                .list_len = 3,
+                .list     = large_file_list}};
+
+        int32_t download_order[128];
+        uint16_t i = 0;
+
+        for ( ; i < 128; ++i )
+        {
+            download_order[i] = i;
+        }
+
+        xi_sft_context_t sft_context;
+        sft_context.updateable_files_download_order = download_order;
+        sft_context.update_current_file             = NULL;
+        sft_context.update_message_fua              = &file_update_available;
+
+
+        xi_state_t state = _xi_sft_select_next_resource_to_download( &sft_context );
+        tt_want_int_op( XI_STATE_OK, ==, state );
+        tt_want_ptr_op( large_file_list, ==, sft_context.update_current_file );
+    } )
+
+XI_TT_TESTCASE_WITH_SETUP(
+    xi_utest__select_next_resource_to_download__large_number_of_files_last_index_is_last_file,
+    xi_utest_setup_basic,
+    xi_utest_teardown_basic,
+    NULL,
+    {
+        // ARRANGE
+        xi_control_message_file_desc_ext_t large_file_list[128];
+        memset( large_file_list, 0, 128 * sizeof( xi_control_message_file_desc_ext_t ) );
+
+        xi_control_message_t file_update_available = {
+            .file_update_available = {
+                .common = {.msgtype = XI_CONTROL_MESSAGE_SC__SFT_FILE_UPDATE_AVAILABLE,
+                           .msgver  = 1},
+                .list_len = 128,
+                .list     = large_file_list}};
+
+        int32_t download_order[128];
+        uint16_t i = 0;
+
+        for ( ; i < 127; ++i )
+        {
+            download_order[i] = -1;
+        }
+        download_order[127] = 127;
+
+        xi_sft_context_t sft_context;
+        sft_context.updateable_files_download_order = download_order;
+        sft_context.update_current_file             = NULL;
+        sft_context.update_message_fua              = &file_update_available;
+
+        xi_state_t state = _xi_sft_select_next_resource_to_download( &sft_context );
+        tt_want_int_op( XI_STATE_OK, ==, state );
+        tt_want_ptr_op( sft_context.update_current_file, ==, &large_file_list[127] );
+    } )
+
+XI_TT_TESTCASE_WITH_SETUP(
+    xi_utest__select_next_resource_to_download__large_number_of_files_last_index_is_first_file,
+    xi_utest_setup_basic,
+    xi_utest_teardown_basic,
+    NULL,
+    {
+        // ARRANGE
+        xi_control_message_file_desc_ext_t large_file_list[128];
+        memset( large_file_list, 0, 128 * sizeof( xi_control_message_file_desc_ext_t ) );
+
+        xi_control_message_t file_update_available = {
+            .file_update_available = {
+                .common = {.msgtype = XI_CONTROL_MESSAGE_SC__SFT_FILE_UPDATE_AVAILABLE,
+                           .msgver  = 1},
+                .list_len = 128,
+                .list     = large_file_list}};
+
+        int32_t download_order[128];
+        uint16_t i = 0;
+
+        for ( ; i < 127; ++i )
+        {
+            download_order[i] = -1;
```

Comment:
I would add few more unit tests with defining download order in the following ways:
```
download_order[i] = 127 - i;
download_order[i] = (i < 64) ? 126 - 2 * i : 2 * i - 127; // this is a V, zero at i==63, left side even, right side odd numbers, pick order should be: 63, 64, 62, 65, 61, 66, 60, 67, ... , 2, 125, 1, 126, 0, 127 
download_order[i] = (i < 64) ? 126 - 2 * i : 255 - 2 * i); // this is two backslashes, 126 at i==0, 0 at i==63, 127 at i==64, 1 at i==127, expected order: 63, 127, 62, 126, 61, 125, 60, 124, ... , 1, 65, 0, 64
``` 
I would do multiple `_xi_sft_select_next_resource_to_download` calls in the last unit tests and check all return values of the `_xi_sft_select_next_resource_to_download` calls to the end. 128 calls.